components/ledger/ledger-utxo-flow/src/main/kotlin/net/corda/ledger/utxo/flow/impl/flows/finality/v1/UtxoReceiveFinalityFlowV1.kt

```diff
@@ -73,17 +76,39 @@ class UtxoReceiveFinalityFlowV1(
             session.send(payload)
             throw CordaRuntimeException(payload.message)
         }
-        transaction = receiveSignaturesAndAddToTransaction(transaction)
-        verifyAllReceivedSignatures(transaction)
-        persistenceService.persist(transaction, TransactionStatus.UNVERIFIED)
+
+        transaction = receiveAndPersistSignaturesOrSkip(transaction, shouldWaitForAdditionalSignatures)
         transaction = receiveNotarySignaturesAndAddToTransaction(transaction)
         persistNotarizedTransaction(transaction)
         return transaction
     }
 
     @Suspendable
-    private fun receiveTransactionAndBackchain(): UtxoSignedTransactionInternal {
-        val initialTransaction = session.receive(UtxoSignedTransactionInternal::class.java)
+    private fun receiveAndPersistSignaturesOrSkip(
+        transaction: UtxoSignedTransactionInternal,
+        shouldWaitForAdditionalSignatures: Boolean
+    ): UtxoSignedTransactionInternal {
+        var verifiedTransaction = transaction
+        if (shouldWaitForAdditionalSignatures) {
+            verifiedTransaction = receiveSignaturesAndAddToTransaction(verifiedTransaction)
+            verifyAllReceivedSignatures(verifiedTransaction)
+            persistenceService.persist(verifiedTransaction, TransactionStatus.UNVERIFIED)
+        } else {
+            verifyAllReceivedSignatures(verifiedTransaction)
+        }
+
+        return verifiedTransaction
```

Comment:
It is better to avoid `verify` in this case. (It is used for quite a few things already. The base class has 5 verifySomething methods.)

`transactionWithAllSignatorySignatures` would be the correct name. But it is pretty ugly... :)

Maybe just "kinda" inlining it?
```suggestion
        return if (shouldWaitForAdditionalSignatures) {
            receiveSignaturesAndAddToTransaction(transaction).also {
              verifyAllReceivedSignatures(it)
              persistenceService.persist(it, TransactionStatus.UNVERIFIED)
            }
        } else {
            verifyAllReceivedSignatures(transaction)
            transaction
        }
```